### ISSUE TITLE
missing pixels when using the iiif retiler on a dzi image

### ISSUE THREAD

**Issue by lovasoa** [lovasoa/dezoomify-rs#75]
_Originally posted by @Linden6 in https://github.com/lovasoa/dezoomify-rs/pull/74#issuecomment-691531902_

I've just run `dezoomify-rs file.html image.iiif` a couple of times at different zoom levels.

It now works perfectly with an absolute URL in a local http file.

And pixels now cover **nearly** all the source tiles.  However when calling the first Deep Zoom source from 
https://openseadragon.github.io/examples/tilesource-dzi/ at zoom level 2 (1757 x  2306 pixels) via:
```<!doctype html>
<html>
  <head>
    <meta charset="utf-8">
    <title>Image Viewer</title>
    <style>
      #viewer {
	  min-height: 300px;
	  height: 95vh;
      }
    </style>
    <script src="https://openseadragon.github.io/openseadragon/openseadragon.min.js"></script>
  </head>
  <body>

    <div id="viewer"></div>

    <script>
      OpenSeadragon({
	  id:              "viewer",
	  prefixUrl:       "https://openseadragon.github.io/openseadragon/images/",
	  tileSources: {
              Image: {
		  xmlns:    "http://schemas.microsoft.com/deepzoom/2008",
		  Url:      "https://openseadragon.github.io/example-images/highsmith/highsmith_files/",
		  Format:   "jpg", 
		  Overlap:  "2", 
		  TileSize: "256",
		  Size: {
                      Height: "9221",
                      Width:  "7026"
		  }
              }
	  }
      });
    </script>
  </body>
</html>
```
The following two warnings were received:

> [2020-09-12T17:53:24Z WARN  dezoomify_rs::encoder::retiler] The target tile of size x=1757 y=2048 at zoom level 4 and position x=0 y=0 was not fully covered by source tiles. It misses 512 pixels.
> [2020-09-12T17:53:25Z WARN  dezoomify_rs::encoder::retiler] The target tile of size x=1757 y=258 at zoom level 4 and position x=0 y=2048 was not fully covered by source tiles. It misses 504 pixels.

With the two most zoomed out tiles:
./0,0,1757,2048/440,512/0/default.jpg  - having a 1px right black edge
./0,2048,1757,258/440,65/0/default.jpg - having a 1px right and bottom black edge

And when calling the same command and choosing zoom level 0, the following warning messages were received:

> [2020-09-12T18:41:13Z WARN  dezoomify_rs::encoder::retiler] The target tile of size x=882 y=2048 at zoom level 4 and position x=6144 y=6144 was not fully covered by source tiles. It misses 512 pixels.
> [2020-09-12T18:41:13Z WARN  dezoomify_rs::encoder::retiler] The target tile of size x=2048 y=1029 at zoom level 4 and position x=4096 y=8192 was not fully covered by source tiles. It misses 512 pixels.
> [2020-09-12T18:41:13Z WARN  dezoomify_rs::encoder::retiler] The target tile of size x=882 y=2048 at zoom level 4 and position x=6144 y=4096 was not fully covered by source tiles. It misses 512 pixels.
> [2020-09-12T18:41:13Z WARN  dezoomify_rs::encoder::retiler] The target tile of size x=2048 y=1029 at zoom level 4 and position x=2048 y=8192 was not fully covered by source tiles. It misses 512 pixels.
> [2020-09-12T18:41:13Z WARN  dezoomify_rs::encoder::retiler] The target tile of size x=882 y=1029 at zoom level 4 and position x=6144 y=8192 was not fully covered by source tiles. It misses 478 pixels.
> [2020-09-12T18:41:14Z WARN  dezoomify_rs::encoder::retiler] The target tile of size x=2048 y=1029 at zoom level 4 and position x=0 y=8192 was not fully covered by source tiles. It misses 512 pixels.
> [2020-09-12T18:41:14Z WARN  dezoomify_rs::encoder::retiler] The target tile of size x=882 y=2048 at zoom level 4 and position x=6144 y=2048 was not fully covered by source tiles. It misses 512 pixels.
> [2020-09-12T18:41:14Z WARN  dezoomify_rs::encoder::retiler] The target tile of size x=882 y=2048 at zoom level 4 and position x=6144 y=0 was not fully covered by source tiles. It misses 512 pixels.
> [2020-09-12T18:41:14Z WARN  dezoomify_rs::encoder::retiler] The target tile of size x=2930 y=4096 at zoom level 8 and position x=4096 y=0 was not fully covered by source tiles. It misses 512 pixels.
> [2020-09-12T18:41:14Z WARN  dezoomify_rs::encoder::retiler] The target tile of size x=4096 y=1029 at zoom level 8 and position x=0 y=8192 was not fully covered by source tiles. It misses 512 pixels.
> [2020-09-12T18:41:14Z WARN  dezoomify_rs::encoder::retiler] The target tile of size x=2930 y=1029 at zoom level 8 and position x=4096 y=8192 was not fully covered by source tiles. It misses 495 pixels.
> [2020-09-12T18:41:15Z WARN  dezoomify_rs::encoder::retiler] The target tile of size x=2930 y=4096 at zoom level 8 and position x=4096 y=4096 was not fully covered by source tiles. It misses 512 pixels.
> [2020-09-12T18:41:15Z WARN  dezoomify_rs::encoder::retiler] The target tile of size x=7026 y=8192 at zoom level 16 and position x=0 y=0 was not fully covered by source tiles. It misses 512 pixels.
> [2020-09-12T18:41:15Z WARN  dezoomify_rs::encoder::retiler] The target tile of size x=7026 y=1029 at zoom level 16 and position x=0 y=8192 was not fully covered by source tiles. It misses 504 pixels.

Again these tiles are edge tiles with edges along the right or bottom of the image, and the 1px black edges are present.

The overall dimensions of the tiles appear correct, so I'm not sure it's a rounding error this time.

Thanks!

**Comment by Linden6** [comment]
_'Every level passes the source tile to it's child when it is done with it.'_ - from retiler.rs line 36.

At the highest zoom level (warning message level 1) the source tiles are laid out with an overlap of 2px:
![overlap1](https://user-images.githubusercontent.com/43049925/93006383-75764f80-f553-11ea-985a-bac9b057591d.jpg)

As we zoom up to the next level (warning message level 2), the source tiles are scaled so that they are now half as wide and half as high, their overlap is now 1px: 
![overlap2](https://user-images.githubusercontent.com/43049925/93006385-7ad39a00-f553-11ea-90c5-4d4129bad92b.jpg)

As we zoom up to the next level (warning message level 4), the source tiles are scaled so that they are now one quarter as wide and one quarter as high, their overlap is now 0.5px:
![overlap3](https://user-images.githubusercontent.com/43049925/93006389-7f984e00-f553-11ea-9dad-9351d02b3fc6.jpg)

But with an overlap of 0.5px is not possible to align neighbouring tiles.  I think this is why pixels start disappearing and warning messages start appearing at level 4 and above.


If this is the problem then the easiest solution is probably just to trim the 'overlap pixels' off tiles at source and carry out the same operation as for used for the IIIF source tiles where no missing pixel problems are encountered.

Of course as I haven't got my head round the actual method currently employed - so this may all be nonsense!

**Comment by lovasoa** [comment]
Yes, my resizing method has an issue. I first compute the top-left position of the tile in the target level using `floor(original_position / scale_factor)` (which is correct), but then, I compute the target size using `ceil(original_ceil / scale_factor)`. What we actually want to make sure all pixels are covered is computing the coordinates of the bottom right pixel using `ceil(original_bottom_right / scale_factor)`, and then compute the size based on this.

For instance, in this situation 

<img width="762" alt="image" src="https://user-images.githubusercontent.com/552629/93017138-988e1700-f5c6-11ea-8113-48fe1c017487.png">



we want the source tile too cover all target pixels, whereas in the current configuration, it will cover only the first one.

**Comment by lovasoa** [comment]
Tha's fixed, you can try build 7b24fbe83e5cadee671d02222fa45797cc1042af

**Comment by Linden6** [comment]
I've just tested it and it's perfect now.  Many thanks for all your hard work!
